src/providers/assets/operatorsProvider.tsx

```diff
@@ -0,0 +1,209 @@
+
+import {
+  GetDistributionBucketsWithOperatorsDocument,
+  GetDistributionBucketsWithOperatorsQuery,
+  GetDistributionBucketsWithOperatorsQueryVariables,
+  GetStorageBucketsDocument,
+  GetStorageBucketsQuery,
+  GetStorageBucketsQueryVariables,
+} from '@/api/queries'
+import { ViewErrorFallback } from '@/components/ViewErrorFallback'
+import { ConsoleLogger, SentryLogger } from '@/utils/logs'
+import { getRandomIntInclusive } from '@/utils/number'
+
+import { OperatorInfo } from './types'
+
+type BagOperatorsMapping = Record<string, OperatorInfo[]>
+
+type OperatorsContextValue = {
+  distributionOperatorsMappingPromiseRef: React.MutableRefObject<Promise<BagOperatorsMapping> | undefined>
+  storageOperatorsMappingPromiseRef: React.MutableRefObject<Promise<BagOperatorsMapping> | undefined>
+  failedStorageOperatorIds: string[]
+  setFailedStorageOperatorIds: React.Dispatch<SetStateAction<string[]>>
+}
+const OperatorsContext = React.createContext<OperatorsContextValue | undefined>(undefined)
+OperatorsContext.displayName = 'OperatorsContext'
+
+export const OperatorsContextProvider: React.FC = ({ children }) => {
+  const distributionOperatorsMappingPromiseRef = useRef<Promise<BagOperatorsMapping>>()
+  const storageOperatorsMappingPromiseRef = useRef<Promise<BagOperatorsMapping>>()
+  const [distributionOperatorsError, setDistributionOperatorsError] = useState<unknown>(null)
+  const [storageOperatorsError, setStorageOperatorsError] = useState<unknown>(null)
+  const [failedStorageOperatorIds, setFailedStorageOperatorIds] = useState<string[]>([])
+
+  const client = useApolloClient()
+
+  // runs once - fetch all distribution operators and create associated mappings
+  useEffect(() => {
+    const distributionOperatorsPromise = client.query<
+      GetDistributionBucketsWithOperatorsQuery,
+      GetDistributionBucketsWithOperatorsQueryVariables
+    >({
+      query: GetDistributionBucketsWithOperatorsDocument,
+      fetchPolicy: 'network-only',
+    })
+    distributionOperatorsMappingPromiseRef.current = distributionOperatorsPromise.then((result) => {
+      const mapping: BagOperatorsMapping = {}
+      const buckets = result.data.distributionBuckets
+      buckets.forEach((bucket) => {
+        const bagIds = bucket.bags.map((bag) => bag.id)
+
+        // we need to filter operators manually as query node doesn't support filtering this deep
+        const operatorsInfos: OperatorInfo[] = bucket.operators
+          .filter((operator) => operator.metadata?.nodeEndpoint?.includes('http') && operator.status === 'ACTIVE')
+          .map((operator) => ({ id: operator.id, endpoint: operator.metadata?.nodeEndpoint || '' }))
+
+        bagIds.forEach((bagId) => {
+          if (!mapping[bagId]) {
+            mapping[bagId] = operatorsInfos
+          } else {
+            // TODO: make sure endpoints are unique
+            mapping[bagId] = [...mapping[bagId], ...operatorsInfos]
```

Comment:
Is this still needed before merge?

Reply:
No, we can leave that for cleanup as part of #1718